File: backend/app/models/community.py
"""
FitTracker Backend - 社区相关数据模型
"""

from sqlalchemy import Column, Integer, String, Text, Boolean, DateTime, ForeignKey, JSON, Float
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.core.database import Base

class Post(Base):
    """社区动态表"""
    __tablename__ = "posts"

    id = Column(Integer, primary_key=True, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    deleted_at = Column(DateTime(timezone=True))
    
    # 用户关系
    user_id = Column(String, ForeignKey("users.id"), nullable=False)
    
    # 内容信息
    content = Column(Text, nullable=False)
    images = Column(JSON, default=list)  # 图片URL列表
    video_url = Column(String)  # 视频URL
    
    # 动态类型
    post_type = Column(String, default="general")  # general, workout, nutrition, education
    is_public = Column(Boolean, default=True)
    
    # 社区扩展
    tags = Column(JSON, default=list)  # 标签列表
    location = Column(String)  # 位置信息
    workout_data = Column(JSON)  # 训练数据
    
    # 状态管理
    is_featured = Column(Boolean, default=False)  # 是否精选
    is_pinned = Column(Boolean, default=False)  # 是否置顶
    
    # 统计信息
    view_count = Column(Integer, default=0)
    share_count = Column(Integer, default=0)
    likes_count = Column(Integer, default=0)
    comments_count = Column(Integer, default=0)
    
    # 约束
    __table_args__ = (
        {"extend_existing": True}
    )

class Comment(Base):
    """评论表"""
    __tablename__ = "comments"

    id = Column(Integer, primary_key=True, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    deleted_at = Column(DateTime(timezone=True))
    
    # 关系
    post_id = Column(Integer, ForeignKey("posts.id"), nullable=False)
    user_id = Column(String, ForeignKey("users.id"), nullable=False)
    parent_id = Column(Integer, ForeignKey("comments.id"))  # 父评论ID，支持回复
    
    # 内容
    content = Column(Text, nullable=False)
    
    # 统计
    likes_count = Column(Integer, default=0)
    
    # 约束
    __table_args__ = (
        {"extend_existing": True}
    )

class Like(Base):
    """点赞表"""
    __tablename__ = "likes"

    id = Column(Integer, primary_key=True, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # 关系
    user_id = Column(String, ForeignKey("users.id"), nullable=False)
    post_id = Column(Integer, ForeignKey("posts.id"), nullable=True)
    comment_id = Column(Integer, ForeignKey("comments.id"), nullable=True)
    
    # 约束：只能点赞动态或评论中的一个
    __table_args__ = (
        {"extend_existing": True}
    )

class Follow(Base):
    """关注关系表"""
    __tablename__ = "follows"

    id = Column(Integer, primary_key=True, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # 关系
    follower_id = Column(String, ForeignKey("users.id"), nullable=False)  # 关注者
    following_id = Column(String, ForeignKey("users.id"), nullable=False)  # 被关注者
    
    # 约束
    __table_args__ = (
        {"extend_existing": True}
    )

class Topic(Base):
    """话题表"""
    __tablename__ = "topics"

    id = Column(Integer, primary_key=True, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # 基本信息
    name = Column(String, nullable=False, unique=True)
    description = Column(Text)
    icon = Column(String)  # 话题图标
    
    # 统计
    posts_count = Column(Integer, default=0)
    followers_count = Column(Integer, default=0)
    
    # 状态
    is_active = Column(Boolean, default=True)
    
    # 约束
    __table_args__ = (
        {"extend_existing": True}
    )

class PostTopic(Base):
    """动态话题关联表"""
    __tablename__ = "post_topics"

    id = Column(Integer, primary_key=True, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # 关系
    post_id = Column(Integer, ForeignKey("posts.id"), nullable=False)
    topic_id = Column(Integer, ForeignKey("topics.id"), nullable=False)
    
    # 约束
    __table_args__ = (
        {"extend_existing": True}
    )

class Report(Base):
    """举报表"""
    __tablename__ = "reports"

    id = Column(Integer, primary_key=True, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # 关系
    reporter_id = Column(String, ForeignKey("users.id"), nullable=False)  # 举报人
    post_id = Column(Integer, ForeignKey("posts.id"), nullable=True)
    comment_id = Column(Integer, ForeignKey("comments.id"), nullable=True)
    user_id = Column(String, ForeignKey("users.id"), nullable=True)  # 被举报用户
    
    # 举报信息
    reason = Column(String, nullable=False)  # 举报原因
    description = Column(Text)  # 详细描述
    
    # 状态
    status = Column(String, default="pending")  # pending, processed, rejected
    processed_at = Column(DateTime(timezone=True))
    processed_by = Column(String, ForeignKey("users.id"))
    
    # 约束
    __table_args__ = (
        {"extend_existing": True}
    )

File: backend-go/internal/models/community.go
package models

import (
	"time"
)

// Post 社区动态模型
type Post struct {
	ID           string       `json:"id" gorm:"primaryKey"`
	UserID       string       `json:"user_id" gorm:"type:uuid;not null"`
	Content      string       `json:"content" gorm:"not null"`
	Type         string       `json:"type"` // text, image, video, workout, checkin
	Images       []string     `json:"images" gorm:"type:jsonb"`
	VideoURL     string       `json:"video_url"`
	Tags         []string     `json:"tags" gorm:"type:jsonb"`
	Location     string       `json:"location"`
	WorkoutData  *WorkoutData `json:"workout_data" gorm:"type:jsonb"`
	LikeCount    int          `json:"like_count"`
	CommentCount int          `json:"comment_count"`
	ShareCount   int          `json:"share_count"`
	IsLiked      bool         `json:"is_liked" gorm:"-"`
	IsFeatured   bool         `json:"is_featured" gorm:"column:is_featured"` // 是否精选
	IsPinned     bool         `json:"is_pinned" gorm:"column:is_pinned"`     // 是否置顶
	CreatedAt    time.Time    `json:"created_at"`
	UpdatedAt    time.Time    `json:"updated_at"`

	// 关联数据
	User     User       `json:"user" gorm:"foreignKey:UserID"`
	Comments []Comment  `json:"comments" gorm:"foreignKey:PostID"`
	Likes    []PostLike `json:"likes" gorm:"foreignKey:PostID"`
}

// TableName 指定表名
func (Post) TableName() string {
	return "posts"
}

// Comment 评论模型
type Comment struct {
	ID        string    `json:"id" gorm:"primaryKey"`
	PostID    string    `json:"post_id" gorm:"not null"`
	UserID    string    `json:"user_id" gorm:"type:uuid;not null"`
	Content   string    `json:"content" gorm:"not null"`
	ParentID  string    `json:"parent_id"` // 回复的评论ID
	LikeCount int       `json:"like_count"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	// 关联数据
	User    User      `json:"user" gorm:"foreignKey:UserID"`
	Replies []Comment `json:"replies" gorm:"foreignKey:ParentID"`
}

// PostLike 点赞模型
type PostLike struct {
	ID        string    `json:"id" gorm:"primaryKey"`
	PostID    string    `json:"post_id" gorm:"not null"`
	UserID    string    `json:"user_id" gorm:"type:uuid;not null"`
	CreatedAt time.Time `json:"created_at"`
}

// Follow 关注关系模型
type Follow struct {
	ID          string    `json:"id" gorm:"primaryKey"`
	FollowerID  string    `json:"follower_id" gorm:"type:uuid;not null"`
	FollowingID string    `json:"following_id" gorm:"type:uuid;not null"`
	CreatedAt   time.Time `json:"created_at"`
}

// Topic 话题模型
type Topic struct {
	ID             string    `json:"id" gorm:"primaryKey"`
	Name           string    `json:"name" gorm:"uniqueIndex;not null"`
	Description    string    `json:"description"`
	ImageURL       string    `json:"image_url"`
	PostCount      int       `json:"post_count"`
	FollowersCount int       `json:"followers_count"`
	IsActive       bool      `json:"is_active"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
}

// PostTopic 动态话题关联模型
type PostTopic struct {
	ID      string `json:"id" gorm:"primaryKey"`
	PostID  string `json:"post_id" gorm:"not null"`
	TopicID string `json:"topic_id" gorm:"not null"`
}

// Report 举报模型
type Report struct {
	ID          string    `json:"id" gorm:"primaryKey"`
	ReporterID  string    `json:"reporter_id" gorm:"not null"`
	PostID      string    `json:"post_id"`
	CommentID   string    `json:"comment_id"`
	UserID      string    `json:"user_id"` // 被举报用户
	Reason      string    `json:"reason" gorm:"not null"`
	Description string    `json:"description"`
	Status      string    `json:"status" gorm:"default:'pending'"` // pending, processed, rejected
	ProcessedAt time.Time `json:"processed_at"`
	ProcessedBy string    `json:"processed_by"`
	CreatedAt   time.Time `json:"created_at"`
}

// CreatePostRequest 创建动态请求
type CreatePostRequest struct {
	Content     string       `json:"content" binding:"required"`
	Type        string       `json:"type" binding:"required"`
	Images      []string     `json:"images"`
	VideoURL    string       `json:"video_url"`
	Tags        []string     `json:"tags"`
	Location    string       `json:"location"`
	WorkoutData *WorkoutData `json:"workout_data"`
}

// UpdatePostRequest 更新动态请求
type UpdatePostRequest struct {
	Content     string       `json:"content"`
	Images      []string     `json:"images"`
	VideoURL    string       `json:"video_url"`
	Tags        []string     `json:"tags"`
	Location    string       `json:"location"`
	WorkoutData *WorkoutData `json:"workout_data"`
}

// CreateCommentRequest 创建评论请求
type CreateCommentRequest struct {
	Content  string `json:"content" binding:"required"`
	ParentID string `json:"parent_id"`
}

// CreateTopicRequest 创建话题请求
type CreateTopicRequest struct {
	Name        string `json:"name" binding:"required"`
	Description string `json:"description"`
	ImageURL    string `json:"image_url"`
}

// CreateReportRequest 创建举报请求
type CreateReportRequest struct {
	PostID      string `json:"post_id"`
	CommentID   string `json:"comment_id"`
	UserID      string `json:"user_id"`
	Reason      string `json:"reason" binding:"required"`
	Description string `json:"description"`
}

// PostResponse 动态响应
type PostResponse struct {
	ID           string       `json:"id"`
	UserID       string       `json:"user_id"`
	Content      string       `json:"content"`
	Type         string       `json:"type"`
	Images       []string     `json:"images"`
	VideoURL     string       `json:"video_url"`
	Tags         []string     `json:"tags"`
	Location     string       `json:"location"`
	WorkoutData  *WorkoutData `json:"workout_data"`
	LikeCount    int          `json:"like_count"`
	CommentCount int          `json:"comment_count"`
	ShareCount   int          `json:"share_count"`
	IsLiked      bool         `json:"is_liked"`
	IsFeatured   bool         `json:"is_featured"`
	IsPinned     bool         `json:"is_pinned"`
	CreatedAt    time.Time    `json:"created_at"`
	UpdatedAt    time.Time    `json:"updated_at"`
	User         User         `json:"user"`
	Comments     []Comment    `json:"comments"`
}

// CommentResponse 评论响应
type CommentResponse struct {
	ID        string            `json:"id"`
	PostID    string            `json:"post_id"`
	UserID    string            `json:"user_id"`
	Content   string            `json:"content"`
	ParentID  string            `json:"parent_id"`
	LikeCount int               `json:"like_count"`
	CreatedAt time.Time         `json:"created_at"`
	UpdatedAt time.Time         `json:"updated_at"`
	User      User              `json:"user"`
	Replies   []CommentResponse `json:"replies"`
}

// TopicResponse 话题响应
type TopicResponse struct {
	ID             string    `json:"id"`
	Name           string    `json:"name"`
	Description    string    `json:"description"`
	ImageURL       string    `json:"image_url"`
	PostCount      int       `json:"post_count"`
	FollowersCount int       `json:"followers_count"`
	IsActive       bool      `json:"is_active"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
}
